README.md
# Encrypt Cloudflare Workers KV

This library provides wrappers on the `put` and `get` functions from the Cloudflare Workers runtime API for writing to and reading from [Workers KV](https://developers.cloudflare.com/workers/reference/apis), encrypting values before `put` and decrypting values after `get`. Encryption is implemented using the [Web Crypto API](https://developers.cloudflare.com/workers/reference/apis/web-crypto/) to derive AES-GCM keys from a password based key (PBKDF2).

By default all data stored in Cloudflare [Workers KV](https://developers.cloudflare.com/workers/reference/storage) is encrypted at rest:

> All values are encrypted at rest with 256-bit AES-GCM, and only decrypted by the process executing your Worker scripts or responding to your API requests. ([docs](https://developers.cloudflare.com/workers/reference/storage))

However, there are a variety of reasons you may want to add your own encryption to the values stored in Workers KV. For example, permissions to access Workers KV are scoped at the account level. For those working in shared team or organizational accounts, this means you cannot limit access to specific KV namespaces, as anyone in that account with access to Workers can read the stored data in all KV namespaces in that account.

## Installation

Use npm to install this package while in the root directory of your Workers project:

`npm i encrypt-workers-kv`

## Usage

<dl>
<dt><a href="#putEncryptedKV">putEncryptedKV(namespace, key, data, password, options)</a></dt>
<dd><p>Wrapper on Workers KV put command that encrypts data prior to storage</p>
</dd>
<dt><a href="#getDecryptedKV">getDecryptedKV(namespace, key, password)</a></dt>
<dd><p>Wrapper on Workers KV get command that decrypts data after getting from storage</p>
</dd>
</dl>

<a name="putEncryptedKV"></a>

### putEncryptedKV(namespace, key, data, password, options)

Wrapper on Workers KV put command that encrypts data prior to storage

| Param      | Type                                 | Description                                                                                                          |
| ---------- | ------------------------------------ | -------------------------------------------------------------------------------------------------------------------- |
| namespace  | <code>KVNamespace</code>             | the binding to the namespace that script references                                                                  |
| key        | <code>string</code>                  | the key in the namespace used to reference the stored value                                                          |
| data       | <code>string</code> or `ArrayBuffer` | the data to encrypt and store in KV                                                                                  |
| password   | <code>string</code>                  | the password to be used to encrypt the data                                                                          |
| iterations | <code>number</code>                  | optional number of iterations used by the PBKDF2 to derive the key. Default 10000                                    |
| options    | <code>Object</code>                  | optional KV put fields ([docs](https://developers.cloudflare.com/workers/reference/apis/kv/#creating-expiring-keys)) |

Returns encrypted value as string - `Promise<ArrayBuffer>`

Sample implementation:

```javascript
let data = await request.text()
try {
  await putEncryptedKV(ENCRYPTED, 'data', data, password)
  return new Response('Secret stored successfully')
} catch (e) {
  return new Response(e.message, { status: e.status })
}
```

<a name="getDecryptedKV"></a>

### getDecryptedKV(namespace, key, password)

Wrapper on Workers KV get command that decrypts data after getting from storage

| Param     | Type                     | Description                                                 |
| --------- | ------------------------ | ----------------------------------------------------------- |
| namespace | <code>KVNamespace</code> | the binding to the namespace that script references         |
| key       | <code>string</code>      | the key in the namespace used to reference the stored value |
| password  | <code>string</code>      | the password used to encrypt the data                       |

Returns decrypted value as string - `Promise<ArrayBuffer>`

Sample implementation:

```javascript
try {
  let decryptedData = await getDecryptedKV(ENCRYPTED, 'data', password)
  let strDecryptedData = dec.decode(decryptedData)
  return new Response(`${strDecryptedData}`, {
    headers: { 'content-type': 'text/html; charset=utf-8' },
  })
} catch (e) {
  return new Response(e.message, { status: e.status })
}
```

## Logic Flow

An explanation of the steps used behind the scenes for encryption and decryption in `src/index.ts`.

**Encryption:**

1. Creates a password based key (PBKDF2) that will be used to derive the AES-GCM key used for encryption / decryption.
1. Creates an AES-GCM key using the PBKDF2 key and a randomized salt value.
1. Encrypts the input data using the AES-GCM key and a randomized initialization vector (iv).
1. The values used for the password, salt, iv for encryption are needed for decryption. Therefore, creates an ArrayBuffer to be stored that includes the salt that was used when creating the password based key (PBKDF2), iv used for creating the AES key, and the encrypted content. The password should remain secret, so recommend storing it as a [Worker Secret](https://developers.cloudflare.com/workers/reference/apis/environment-variables/#secrets).

**Decryption:**

1. Derives the salt, iv, and encrypted data from the ArrayBuffer.
1. Creates a password based key (PBKDF2) that will be used to derive the AES-GCM key used for encryption / decryption. Password must be the same used for encryption and is obtained from the Workers Secret.
1. Creates an AES-GCM key using the PBKDF2 key and the salt from the ArrayBuffer.
1. Decrypts the input data using the AES-GCM key and the iv from the ArrayBuffer.

## Build and Test

A test worker is used to test the library, located in `test-worker/`. Configure `wrangler.toml` in that folder with your account information. Then, create a new Workers KV namespace and add the configuration to wrangler.toml.

`wrangler kv:namespace create "ENCRYPTED"`

Add the password for PBKDF2 as a Workers Secret.

`wrangler secret put PASSWORD`

To deploy the test worker and run the the automated tests, change directory back to the project root directory and:

`npm run test:deploy && npm run test`

## Sample Implementation

A sample implementation is available in the [Workers Azure AD auth example](https://github.com/bradyjoslin/workers-azuread-example), which uses `encrypt-workers-kv` to encrypt session token information stored as values in Workers KV.

## References

Further info on Web Crypto usage:

- [Ernie Turner: Dodging Web Crypto API Landmines | Web Rebels 2018](https://www.youtube.com/watch?v=lbt2_M1hZeg)
- [Web Crypto API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API)
- Example usage in browser [bradyjoslin/webcrypto-example](https://github.com/bradyjoslin/webcrypto-example)


CHANGELOG.md
# Changelog

## 0.0.3

- **Fixes**

  - Fix options handling in `putEncryptedKV` function [dmitryyankowski](https://github.com/dmitryyankowski), [pr/35](https://github.com/bradyjoslin/encrypt-workers-kv/pull/35)

- **Maintenance**

  - Bump prettier from 2.1.1 to 2.1.2 [pull/34](https://github.com/bradyjoslin/encrypt-workers-kv/pull/34)
  - Bump prettier from 2.1.1 to 2.1.2 in /test-worker [pull/33](https://github.com/bradyjoslin/encrypt-workers-kv/pull/33)
  - Bump node-fetch from 2.6.0 to 2.6.1 [pull/32](https://github.com/bradyjoslin/encrypt-workers-kv/pull/32)
  - Bump prettier from 2.1.0 to 2.1.1 [pull/31](https://github.com/bradyjoslin/encrypt-workers-kv/pull/31)
  - Bump prettier from 2.1.0 to 2.1.1 in /test-worker [pull/30](https://github.com/bradyjoslin/encrypt-workers-kv/pull/30)
  - Bump prettier from 2.0.5 to 2.1.0 in /test-worker [pr/29](https://github.com/bradyjoslin/encrypt-workers-kv/pull/29)
  - Bump prettier from 2.0.5 to 2.1.0 [pr/28](https://github.com/bradyjoslin/encrypt-workers-kv/pull/28)
  - Bump ava from 3.11.1 to 3.12.1 [pr/27](https://github.com/bradyjoslin/encrypt-workers-kv/pull/27)
  - Bump ts-loader from 8.0.2 to 8.0.3 in /test-worker [pr/26](https://github.com/bradyjoslin/encrypt-workers-kv/pull/26)
  - Bump typescript from 3.9.7 to 4.0.2 in /test-worker [pr/25](https://github.com/bradyjoslin/encrypt-workers-kv/pull/25)
  - Bump typescript from 3.9.7 to 4.0.2 [pr/24](https://github.com/bradyjoslin/encrypt-workers-kv/pull/24)

## 0.0.2

Initial release 🐣


src/index.ts
import {
  PutError,
  NotFoundError,
  DecryptionError,
  EncryptionError,
} from './errors'

const enc = new TextEncoder()
const dec = new TextDecoder()

/**
 * Wrapper on Workers KV put command that encrypts data prior to storage
 *
 * @param {KVNamespace} namespace the binding to the namespace that script references
 * @param {string} key the key in the namespace used to reference the stored value
 * @param {string | ArrayBuffer} data the data to encrypt and store in KV
 * @param {string} password the password used to encrypt the data
 * @param {number} iterations optional number of iterations used by the PBKDF2 to derive the key.  Default 10000
 * @param {Object} options optional KV put fields
 * @returns {Promise<ArrayBuffer>} a promise for an encrypted value as ArrayBuffer
 * */
async function putEncryptedKV(
  namespace: KVNamespace,
  key: string,
  data: string | ArrayBuffer,
  password: string,
  iterations: number = 10000,
  options?: {
    expiration?: string | number
    expirationTtl?: string | number
  },
): Promise<ArrayBuffer> {
  data = typeof data === 'string' ? enc.encode(data) : data
  let encryptedData
  try {
    encryptedData = await encryptData(data, password, iterations)
  } catch (e) {
    throw e
  }

  try {
    if (options) {
      await namespace.put(key, encryptedData, options)
    } else {
      await namespace.put(key, encryptedData)
    }
    return encryptedData
  } catch (e) {
    throw new PutError(`Error putting value to kv: ${e.message}`)
  }
}

/**
 * Wrapper on Workers KV get command that decrypts data after getting from storage
 *
 * @param {KVNamespace} namespace the binding to the namespace that script references
 * @param {string} key the key in the namespace used to reference the stored value
 * @param {string} password the password used to encrypt the data
 * @returns {Promise<ArrayBuffer>} a promise for a decrypted value as ArrayBuffer
 * */
async function getDecryptedKV(
  namespace: KVNamespace,
  key: string,
  password: string,
): Promise<ArrayBuffer> {
  let kvEncryptedData = await namespace.get(key, 'arrayBuffer')
  if (kvEncryptedData === null) {
    throw new NotFoundError(`could not find ${key} in your namespace`)
  }

  try {
    let decryptedData = await decryptData(kvEncryptedData, password)
    return decryptedData
  } catch (e) {
    throw e
  }
}

const getPasswordKey = (password: string): PromiseLike<CryptoKey> =>
  crypto.subtle.importKey('raw', enc.encode(password), 'PBKDF2', false, [
    'deriveKey',
  ])

const deriveKey = (
  passwordKey: CryptoKey,
  salt: Uint8Array,
  keyUsage: CryptoKey['usages'],
  iterations: number = 10000,
): PromiseLike<CryptoKey> =>
  crypto.subtle.deriveKey(
    {
      name: 'PBKDF2',
      salt: salt,
      iterations: iterations,
      hash: 'SHA-256',
    },
    passwordKey,
    { name: 'AES-GCM', length: 256 },
    false,
    keyUsage,
  )

async function encryptData(
  secretData: ArrayBuffer,
  password: string,
  iterations: number = 10000,
): Promise<ArrayBuffer> {
  try {
    const salt = crypto.getRandomValues(new Uint8Array(16))
    const iv = crypto.getRandomValues(new Uint8Array(12))
    const passwordKey = await getPasswordKey(password)
    const aesKey = await deriveKey(passwordKey, salt, ['encrypt'], iterations)
    const encryptedContent = await crypto.subtle.encrypt(
      {
        name: 'AES-GCM',
        iv: iv,
      },
      aesKey,
      secretData,
    )

    const encryptedContentArr = new Uint8Array(encryptedContent)
    let iterationsArr = new Uint8Array(enc.encode(iterations.toString()))

    let buff = new Uint8Array(
      iterationsArr.byteLength +
        salt.byteLength +
        iv.byteLength +
        encryptedContentArr.byteLength,
    )
    let bytes = 0
    buff.set(iterationsArr, bytes)
    buff.set(salt, (bytes += iterationsArr.byteLength))
    buff.set(iv, (bytes += salt.byteLength))
    buff.set(encryptedContentArr, (bytes += iv.byteLength))

    return buff.buffer
  } catch (e) {
    throw new EncryptionError(`Error encrypting value: ${e.message}`)
  }
}

async function decryptData(
  encryptedData: ArrayBuffer,
  password: string,
): Promise<ArrayBuffer> {
  try {
    const encryptedDataBuff = new Uint8Array(encryptedData)

    let bytes = 0
    const iterations = Number(
      dec.decode(encryptedDataBuff.slice(bytes, (bytes += 5))),
    )

    const salt = new Uint8Array(encryptedDataBuff.slice(bytes, (bytes += 16)))
    const iv = new Uint8Array(encryptedDataBuff.slice(bytes, (bytes += 12)))
    const data = new Uint8Array(encryptedDataBuff.slice(bytes))

    const passwordKey = await getPasswordKey(password)
    const aesKey = await deriveKey(passwordKey, salt, ['decrypt'], iterations)
    const decryptedContent = await crypto.subtle.decrypt(
      {
        name: 'AES-GCM',
        iv: iv,
      },
      aesKey,
      data,
    )
    return decryptedContent
  } catch (e) {
    throw new DecryptionError(`Error decrypting value: ${e.message}`)
  }
}

export { putEncryptedKV, getDecryptedKV }
